_posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md

```diff
@@ -0,0 +1,166 @@
+---
+layout: post
+title: "How to use agent-side information on the puppetserver"
+author: dev_el_ops
+categories:
+  - module
+  - development
+tags:
+  - modules
+  - puppetserver
+  - custom facts
+  - custom resources
+  - deferred
+---
+
+This question was posted on the [Puppet Community Slack](https://slack.puppet.com/) yesterday:
+
+> Is there any way like the below in puppet code we can use:
+> ```
+> $var=‘run comman locally on agent, if $?=0 echo “present”, else echo “absent”’
+> if $var == present {
+>   # my code
+> } else {
+>   # something else
+> }
+> ```
+(by user [IlovPuppet](https://app.slack.com/team/URGFPBS0N), edited for presentation)
+
+From my answer there, I wrote up a more detailed exposition here.
+Based on the specifics of your situation,
+there are several different options to solve this,
+described in detail below:
+
+* a `Deferred` function call
+* using `onlyif` or `unless` on an `exec` resource
+* a custom fact
+* a custom resource
+
+## `Deferred`
+
+**What is it?**
+
+The `Deferred` function computes a value on the agent everytime the catalog is executed and uses the value immediately in the catalog.
+
+**Example**
+
+The [puppet/vault_lookup](https://forge.puppet.com/puppet/vault_lookup) module shows how to use `Deferred` with its `vault_lookup::lookup` function:
+
+```puppet
+$d = Deferred('vault_lookup::lookup', ["secret/test", 'https://vault.hostname:8200'])
+
+node default {
+  notify { example :
+    message => $d
+  }
+}
+```
+(from [voxpupuli/puppet-vault_lookup](https://github.com/voxpupuli/puppet-vault_lookup#usage))
+
+**Use it ...**
+
+* for values only needed locally on a few agents.
```

Comment:
This is a little unclear. Might this wording be better?
* for values specific to one or a few agents, for which managing via Hiera would be cumbersome

Reply:
:+1: